### PR TITLE
Define Cross-Origin-Resource-Policy response header

### LINKED ISSUE #687: Cross-Origin-Resource-Policy (was: From-Origin)
Websites should have an explicit way to restrict any kind of cross-origin load to protect themselves against Spectre attacks. Content such as images, video, and audio may be sensitive and websites may be protected solely by virtue of their network position (inside a firewall), relying on the same-origin policy to protect against exfiltration.

There's a previous proposal from 2012 called the [From-Origin header](https://www.w3.org/TR/from-origin/) that we'd like to resurrect. With it, a server can send a `From-Origin : same` header on responses it wants to protect from Spectre attacks. Here's a snippet from the currently inactive proposal:

> The From-Origin header can be used to restrict embedding of a resource to only certain origins. When used it must match the following ABNF:
> 
> `From-Origin = "From-Origin" ":" #(serialized-origin | "same")`

[Cross-Origin Read Blocking (CORB)](https://github.com/whatwg/fetch/issues/681) automatically protects against Spectre attacks that load cross-origin, cross-type HTML, XML, and JSON resources, and is based on the browser’s ability to distinguish resource types. We think CORB is a good idea. From-Origin would offer servers an opt-in protection beyond CORB.

In addition to the original proposal, we might want to offer servers a way to say cross-origin requests are OK within the same eTLD+1, e.g. the server may want to say that cross-origin subresources from cdn.example.com may be loaded from *.example.com without listing all those origins.

### PR DESCRIPTION
**Pull request by annevk**
This header makes it easier for sites to block unwanted "no-cors" cross-origin requests.

Tests: ...

Fixes #687.

***
<a href="https://whatpr.org/fetch/733.html" title="Last updated on Jun 18, 2018, 12:20 PM GMT (ed96a17)">Preview</a> | <a href="https://whatpr.org/fetch/733/2f3d04d...ed96a17.html" title="Last updated on Jun 18, 2018, 12:20 PM GMT (ed96a17)">Diff</a>

### CODE REVIEW

File: fetch.bs [hunk 1]
```diff
@@ -2437,6 +2437,64 @@ run these steps:</p>
...
+`<dfn export http-header id=http-cross-origin-resource-policy><code>Cross-Origin-Resource-Policy</code></dfn>`
+response <a for=/>header</a> can be used to require checking a <a for=/>request</a>'s
+<a for=request>current url</a>'s <a for=url>origin</a> against a <a for=/>request</a>'s
+<a for=request>origin</a> when <a for=/>request</a>'s <a for=request>mode</a> is
+"<code>no-cors</code>".
+
+<p>Its <a for=header>value</a> <a>ABNF</a>:
+
+<pre>
+Cross-Origin-Resource-Policy     = %x73.61.6D.65 / %x73.61.6D.65.2D.73.69.74.65 ; "same" / "same-site"; case-sensitive</pre>
```
Comment (mikewest) [inline fetch.bs]: Nit: The last semicolon should be a comma (similar to the `origin-or-null` definition).

File: fetch.bs [hunk 2]
```diff
@@ -2437,6 +2437,64 @@ run these steps:</p>
...
+<ol>
+ <li><p>If <var>request</var>'s <a for=request>mode</a> is not "<code>no-cors</code>", then return
+ <b>allowed</b>.
+
+ <li>
+  <p>If <var>request</var>'s <a for=request>origin</a> is <a>same origin</a> with
+  <var>request</var>'s <a for=request>current url</a>'s <a for=url>origin</a>, then return
+  <b>allowed</b>.
+
+  <p class=note>A cross-origin response redirecting to a same or same-site resource with the
```
Comment (mikewest) [inline fetch.bs]: Nit: This is a little unclear. Is the redirect target same-site with the request's origin, or with the cross-origin response?

File: fetch.bs [hunk 3]
```diff
@@ -2437,6 +2437,64 @@ run these steps:</p>
...
+  <p>If <var>policy</var> is `<code>same-site</code>` and neither of the following is true
+
+  <ul class=brief>
+   <li><p><var>request</var>'s <a for=request>origin</a>'s <a for=origin>host</a>
+   <a>is a registrable domain suffix of or is equal to</a> <var>request</var>'s
+   <a for=request>current url</a>'s <a for=url>host</a>
+
+   <li><p><var>request</var>'s <a for=request>current url</a>'s <a for=url>host</a>
+   <a>is a registrable domain suffix of or is equal to</a> <var>request</var>'s
+   <a for=request>origin</a>'s <a for=origin>host</a>
```
Comment (mikewest) [inline fetch.bs]: I think this isn't broad enough. The bidirectional `is a registrable domain suffix of or is equal to` comparison would return false for a request whose origin's host is `subdomain.subdomain.example.com` and whose current url is `notsubdomain.example.com`.

Instead, I think we need to calculate the relevant registrable domain (perhaps by pointing to the (not terribly clear) algorithm on https://publicsuffix.org/list/) for both the request's origin and the request's current URL, and compare those.

Comment (mikewest) [reply fetch.bs]: (And, actually, looking at that algorithm again, I'm not sure it does what it needs to do for `document.domain` either... It doesn't calculate the registrable domain at all, but the public suffix. I'll file a separate bug.)

Comment (mikewest) [reply fetch.bs]: https://github.com/whatwg/html/issues/3711

File: fetch.bs [hunk 4]
```diff
@@ -2437,6 +2437,64 @@ run these steps:</p>
...
+  <p class=note>A cross-origin response redirecting to a same or same-site resource with the
+  `<a http-header><code>Cross-Origin-Resource-Policy</code></a>` header specified does not affect
+  anything.
+  <!-- We could make this have an effect if we fix https://github.com/whatwg/fetch/pull/594 first,
+       but even then we normally do not let this have any effect for "no-cors" so it would be
+       somewhat inconsistent if it did here, but might still be better... -->
+
+ <li><p>Let <var>policy</var> be the <a>combined value</a> with
+ `<a http-header><code>Cross-Origin-Resource-Policy</code></a>` and <var>response</var>'s
+ <a for=response>header list</a>.
```
Comment (mikewest) [inline fetch.bs]: It might be worth adding a note here about the meaning of `same, same` or `same, same-site`, as I believe this means that both will be ignored.

Comment (mikewest) [reply fetch.bs]: I don't think it matters in this patch, but I agree that that's probably what we'd want to do in a future patch (e.g. define this as a [structured header list](https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-04#section-4.2) whose members are strings that we parse as origins or something.

File: fetch.bs (hunk 3 above)
Comment (mikewest) [reply fetch.bs]: Yeah. Ok, assuming I can dig myself out from under everything else, would you prefer those concepts be defined in HTML, Fetch, or URL?

File: fetch.bs (hunk 4 above)
Comment (annevk) [reply fetch.bs]: Hmm, wouldn't that be incompatible with `same` and `same-site` as tokens? And origins cannot be used as tokens, at least not unless https://github.com/httpwg/http-extensions/issues/629 is fixed somehow.

Comment (mikewest) [reply fetch.bs]: I don't think there's any requirement that all the items in a list are the same type. @mnot can confirm, but my understanding is that we could have `same` and `same-site` as "labels" and some other type ("string", I suppose) for origins.

File: fetch.bs [hunk 5]
```diff
@@ -3736,6 +3794,9 @@ Range Requests</cite>. [[HTTP-RANGE]] However, this is not widely supported by b
     </ol>
   </ol>
 
+ <li><p>If the <a>cross-origin resource policy check</a> with <var>request</var> and
```
Comment (annevk) [inline fetch.bs]: @youennf if we fixed #741 and skipped this step if the `CORS flag` is set, we'd address the CORS-preflight issue I think.

File: fetch.bs (hunk 3 above)
Comment (annevk) [reply fetch.bs]: Note: this is now tracked by https://github.com/whatwg/url/pull/391.

**Comment by annevk** [comment]
I pushed a commit that should address the review feedback.

It builds upon https://github.com/whatwg/url/pull/391 (nearly done) and https://github.com/whatwg/fetch/pull/743 (needs review).

@mnot can you confirm that structured headers will allow mixing of identifiers and strings?

**Comment by annevk** [comment]
FYI: I rebased and renamed `same` to `same-origin`.

**Comment by annevk** [comment]
Dependencies have been merged, plan of action is at https://github.com/whatwg/fetch/issues/687#issuecomment-395366076. TL;DR: this will merge next week barring any objections.

File: fetch.bs [hunk 6]
```diff
@@ -2517,6 +2517,71 @@ run these steps:</p>
...
+
+ <li>
+  <p>If the following are true
+
+  <ul class=brief>
+   <li><var>request</var>'s <a for=request>origin</a>'s <a for=url>host</a> is <a>same site</a> with
+   <var>request</var>'s <a for=request>current url</a>'s <a for=url>host</a>
+   <li><var>request</var>'s <a for=request>origin</a>'s <a for=url>scheme</a> is
+   "<code>https</code>" or <var>response</var>'s <a for=response>HTTPS state</a> is
+   "<code>none</code>"
```
Comment (annevk) [inline fetch.bs]: @mikewest is this too cute and should I be checking the scheme of request's current url instead or is this fine/preferred?

Comment (mikewest) [reply fetch.bs]: I feel like I had to think about what this means more than I'd like to. An explicit comparison seems simpler. Or maybe just skipping the check around? Is something like this equivalent?

> If response's HTTPS state is `modern`, request's origin's scheme is `https`.

File: fetch.bs [hunk 7]
```diff
@@ -2517,6 +2517,71 @@ run these steps:</p>
...
+
+ <li>
+  <p>If <var>request</var>'s <a for=request>origin</a> is <a>same origin</a> with
+  <var>request</var>'s <a for=request>current url</a>'s <a for=url>origin</a>, then return
+  <b>allowed</b>.
+
+  <p class="note no-backref">A cross-origin response redirecting to a response that is
+  <a>same origin</a> or <a>same site</a> with the initial request and has a
+  `<a http-header><code>Cross-Origin-Resource-Policy</code></a>` header specified, does not affect
+  anything. I.e., <var>request</var>'s <a for=request>tainted origin flag</a> is not checked.
```
Comment (mikewest) [inline fetch.bs]: Nit: I'd rephrase this a little bit. Perhaps something like "The cross-origin resource policy check does not consider a request's redirect chain when processing a given response's `Cross-Origin-Resource-Policy` header, but only the relationship between the requesting origin and the response which asserts the header."?

File: fetch.bs (hunk 6 above)
Comment (annevk) [reply fetch.bs]: That doesn't seem equivalent as it doesn't clearly evaluate to a boolean.

File: fetch.bs (hunk 7 above)
Comment (annevk) [reply fetch.bs]: Rewrote, but didn't use "redirect chain" as that's not really defined.

**Comment by annevk** [comment]
I filed #767 on adding advice.

**Comment by annevk** [comment]
Bugs:

* https://bugzilla.mozilla.org/show_bug.cgi?id=1459573
* https://bugs.chromium.org/p/chromium/issues/detail?id=853723
* https://bugs.webkit.org/show_bug.cgi?id=186761
* https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/17934554/
